### PR TITLE
Bau interstitial submit confirmation page

### PR DESCRIPTION
**Pull request by MarcUsher**
## 🎫 Ticket
BAU - from pre-launch testing and thinking

## 📝 Description
As part of pre-launch checks, we've decided to add an interstitial page to the submission flow. Adding an extra page to this flow where the user has to again confirm their intent to submit adds a little more friction and should avoid accidental submissions from the view_locked_report or tasklist pages.

This adds the interstitial page and hooks it up to the submission flows for all reports (requiring certification and not requiring certification). It also preserves form validation before a form is submitted or sent for certification, which was previously done by the mark_sent_for_certification() and submit() methods called from the tasklist, so we still call the full form validation instead so we can append errors to the tasklist page and prevent the user from continuing until the errors are fixed.

_NB. Stil to add:_
- _more tests for the updates to the tasklist page flow_
- _a check for reports that don't require certification to prevent circumventing full-form navigation and URL hacking straight to the new interstitial confirm submission page which could throw full form validation errors but not catch/handle them properly_

### Before
image

image

### After
image

image

image

## 🧪 Testing
- please try to break this 🙏😭

- try to submit reports that need certification and don't need certification. Interstitial page should appear for both, with updated button text on tasklist and view_locked_report to take you there (reports requiring certification tasklist button should remain unchanged and still mark the report as sent for certification and lock it). 

### Review Readiness
- [X] PR title and description provide sufficient context for reviewers
- [X] Commits are logical, self-contained, and have clear descriptions

### Testing
- [X] I have tested this change and it meets the acceptance criteria for the ticket
- [X] I need the reviewer(s) to pull and run this change locally
- [X] New (non-developer) functionality has appropriate unit and integration tests
- ~[ ] End-to-end tests have been updated (if applicable)~
- [X] Edge cases and error conditions are tested

### CODE REVIEW

File: tests/integration/access_grant_funding/routes/test_reports.py [hunk 1]
```diff
@@ -547,6 +445,268 @@ def test_decline_sign_off_get(
...
+        assert response.location == url_for(
+            "access_grant_funding.submitted_confirmation",
+            organisation_id=organisation.id,
+            grant_id=grant.id,
+            submission_id=submission_ready_to_submit.id,
+        )
+
+        assert helper.status == SubmissionStatusEnum.SUBMITTED
+        assert helper.events.submission_state.is_submitted
+        assert len(mock_notification_service_calls) == 3
```
[the remark this answers is not in the document]
Comment (MarcUsher) [reply tests/integration/access_grant_funding/routes/test_reports.py]: More tests added

File: app/access_grant_funding/templates/access_grant_funding/reports/submit_report.html [hunk 2]
```diff
@@ -0,0 +1,36 @@
+{% extends "access_grant_funding/base.html" %}
+{% from "govuk_frontend_jinja/components/back-link/macro.html" import govukBackLink %}
+
+{% set page_title = "Confirm and submit report: " + submission_helper.name %}
+
+{% set active_item_identifier = "reports" %}
+
+{% set return_link = url_for("access_grant_funding.route_to_submission", organisation_id=grant_recipient.organisation.id, grant_id=grant_recipient.grant.id, collection_id = submission_helper.collection_id) %}
```
[the remark this answers is not in the document]
Comment (MarcUsher) [reply app/access_grant_funding/templates/access_grant_funding/reports/submit_report.html]: This is intentional - if you come to this page when a report requires certification you'll be coming from the view_locked_report page, and if you come to this page when a report doesn't require certification you'll be coming form the tasklist. `Back` and `Cancel` should both stop the current journey and send you back to where you came from, and the `route_to_submission` route already has the logic for dealing with both report certification states.

File: app/access_grant_funding/routes/reports.py [hunk 3]
```diff
@@ -224,6 +222,73 @@ def decline_report(
...
+        except SubmissionValidationFailed as e:
+            flash(e.error_message, FlashMessageType.SUBMISSION_VALIDATION_ERROR)
+            return redirect(
+                url_for(
+                    "access_grant_funding.route_to_submission",
+                    organisation_id=organisation_id,
+                    grant_id=grant_id,
+                    collection_id=submission_helper.collection_id,
+                )
+            )
```
[the remark this answers is not in the document]
Comment (MarcUsher) [reply app/access_grant_funding/routes/reports.py]: This is intentional (though maybe brittle) - we'll only ever be in a state where someone could URL hack and submit a submission with validation errors for reports that don't require certification - any reports that do need certification can only get to this page if the submission is awaiting_sign_off, and they can only be awaiting_sign_off if the button was clicked on the tasklist (where this validation also runs and is caught). 

But this is a lot of assuming and not ideal 😬

File: app/access_grant_funding/routes/reports.py [hunk 4]
```diff
@@ -224,6 +219,118 @@ def decline_report(
...
+        )
+        return redirect(
+            url_for("access_grant_funding.list_reports", organisation_id=organisation_id, grant_id=grant_id)
+        )
+
+    form = GenericSubmitForm()
+
+    if form.validate_on_submit():
+        try:
+            submission_helper.submit(user)
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/access_grant_funding/routes/reports.py]: 🤔🤔🤔 feels wrong

File: app/access_grant_funding/routes/reports.py [hunk 5]
```diff
@@ -224,6 +219,118 @@ def decline_report(
...
+    organisation_id: UUID, grant_id: UUID, submission_id: UUID
+) -> ResponseReturnValue:
+    grant_recipient = get_grant_recipient(grant_id, organisation_id)
+    submission_helper = SubmissionHelper.load(submission_id=submission_id, grant_recipient_id=grant_recipient.id)
+    user = get_current_user()
+
+    if not submission_helper.is_awaiting_sign_off:
+        current_app.logger.warning(
+            "Confirm certify and submit loaded incorrectly by %(user_id)s for submission %(submission_id)s",
+            {"user_id": user.id, "submission_id": submission_id},
```
Comment (samuelhwilliams) [inline app/access_grant_funding/routes/reports.py]: extra= here as well I think 🙇

File: app/access_grant_funding/routes/reports.py [hunk 6]
```diff
@@ -224,6 +219,118 @@ def decline_report(
...
+            return redirect(
+                url_for(
+                    "access_grant_funding.submitted_confirmation",
+                    organisation_id=organisation_id,
+                    grant_id=grant_id,
+                    submission_id=submission_id,
+                )
+            )
+        except SubmissionValidationFailed as e:
+            flash(e.error_message, FlashMessageType.SUBMISSION_VALIDATION_ERROR)
```
Comment (samuelhwilliams) [inline app/access_grant_funding/routes/reports.py]: I think I need to do this tech debt as a little self-gift for us 🫣

File: tests/integration/access_grant_funding/routes/test_reports.py [hunk 7]
```diff
@@ -547,6 +453,355 @@ def test_decline_sign_off_get(
...
+class TestConfirmReportSubmission:
+    @pytest.mark.parametrize(
+        "client_fixture, can_access",
+        (
+            ("authenticated_no_role_client", False),
+            ("authenticated_grant_recipient_member_client", False),
+            ("authenticated_grant_recipient_data_provider_client", True),
+            ("authenticated_grant_recipient_certifier_client", False),
+        ),
+    )
```
Comment (samuelhwilliams) [inline tests/integration/access_grant_funding/routes/test_reports.py]: One small thing I tried to do at some point, is define this parametrization as a fixture itself - so that the list of clients+expectations is in one place. Makes it easier to update all tests when we add a new client, which I think we did recently.

Unsure if this is definitely a good idea - each new client then adds 10s/100s of new tests, which maybe is overkill. But does ensure more reliable checking at least? One to momentarily think about here perhaps.

**Review by samuelhwilliams — APPROVED** [review]
nice test coverage and this all works functionally from testing locally. This feels like a good bit of extra friction/protection for users from submitting accidentally.
